catalystneuro/nwb-conversion-tools
# Add buffering to GenericDataChunkIterator

## Motivation

The current version of the GenericDataChunkIterator does initial setup of buffer sizes, and fully integrated chunk specification compatible with `H5DataIO`, but I forgot to set up actual buffering of data in the workflow.


### Optimization Details
This PR adds that functionality in and it was a tad tricky to optimize.

One important thing I've found is that the action of filling a buffer must be somewhat slim even though it gets called less often than the chunk extraction; it can also depend on the actual format of the data storage, too, but we leave that to downstream users when defining `_get_data` and possibly an override of default `buffer_shape` for their own use case.

The most important part, however, is an absolute minimization of the slice calculations going into chunk extraction. Over the last week I was able to improve performance by up to 10% by removing the intermediary index mapping and having a more sophisticated selection generator do it automatically, as well as partitioning the 'inner' buffers (not on a boundary of any axis) to not have to do the safe edge check via calls to `min` on the upper bound. This could actually be improved even further, which I've left as a to-do comment in the code, but calculating the binomial conditions for dimensionality > 2 would take me quite a bit more time and headache to figure out.


### Performance
I've started using DANDIHub to do all my performance benchmarking since I found a great deal of variability on my local system due to bottlenecks from external drivers that host the data (since my local drives are too small to host all the data I work with). Also, it's more standardized for everyone else and it can go beyond any local system in terms of CPU/RAM potential.

Anyway, some results on performance comparison against the old iterator. For a SpikeGLX file stubbed to 5 minutes (~7 GB), and a range of buffer sizes. `buffer_gb` is the key of each dictionary below; keep in mind the actual `buffer_shape` differs by the implementation necessity of each version), with time to complete the `.run_conversion()` step, in seconds, as the corresponding value.

```
v1: {1.0: 203.31578636169434, 1.5: 214.9526071548462, 2.0: 209.98178553581238, 2.5: 213.40778517723083, 3.0: 209.79964518547058, 3.5: 202.3452444076538, 4.0: 202.33232140541077, 4.5: 206.3068766593933, 5.0: 201.93103194236755, 5.5: 202.6948413848877, 6.0: 202.98750615119934, 6.5: 205.2848436832428}
v2: {1.0: 194.65343022346497, 1.5: 177.6259937286377, 2.0: 177.52156281471252, 2.5: 175.40290784835815, 3.0: 176.41599702835083, 3.5: 174.17961430549622, 4.0: 169.06101536750793, 4.5: 168.33040928840637, 5.0: 166.36603903770447, 5.5: 165.263188123703, 6.0: 166.89317774772644, 6.5: 165.98614692687988}
```

As you can see, in the high end there can be a total 20% relative speed improvement against the original iterator, in addition to the full control over chunk (v1 is at the whim of the H5DataIO for chunk shape, through a user can technically control that) and buffer (v1 has little control over this, see below) shapes.


### Debug incorrect/incompatible arguments in `write_recording`
Now, this PR also changes the argument passing of `iterator_options` for `v1` in `write_recording` to allow the user full specification rather than trying to map the meaning of `buffer_gb` onto the true `buffer_size`; after looking deeper into the code for the old `DataChunkIterator`, I realized that the `buffer_size` argument has entirely different behavior depending on (a) the joint argument of `iter_axis`, and (b) the actual type class of the data being accessed. For example, an h5 dataset will buffer an entire axis over a slice defined by the `buffer_size` (meaning there, 'number of data elements to read in along a single fixed axis' but if it's an `array` type (including `memmap`) it will iteratively concatenate multiple entire axes by stacking all channels for a single frame of a recording, or vice versa (meaning there, 'number of a specific orientation of data axes to splice together').


## How to test the behavior?
```
pytest
```

It's still the same iterator used in all testing infrastructure, if anything I need to add some tests to explicitly ensure the old `"v1"` iterator also always works with the interfaces and `write_recording` function.

Yes, I should still add some actual unit tests for the Iterator specifically, but that will be a part of the upstream merge with hdmf anyway so I'll mainly do them over there instead of over here. Mostly wanted to get this merged over here first so we could test it with the existing instances of `RecordingDataChunkIterator` and the `MovieDataChunkIterator` in development to see how well it works on actual data.

## Checklist
